### PR TITLE
UCT/IB: Use DC half-handshake with AR, when RDMA_WRITE is not used

### PR DESCRIPTION
**Pull request by gleon99**
## What
Use DC half-handshake with adaptive routing, when RDMA_WRITE is not used (if RNDV is only RDMA_READ, and no GPU memory).

## Why ?
Following https://github.com/openucx/ucx/pull/7511, we want to improve the logic by allowing DC half-handshake even with AR, in case RDMA_WRITE is not needed.

## How ?
Introduce "features" in UCT iface layer, which are filled by the UCP worker according to its features & config, before calling `uct_iface_open`.

Then, supported iface caps are limited according to the requested features. The UCP layer knows in advance, whether RDMA_WRITE will be needed, and is able to pass that information to the UCT layer.

### CODE REVIEW

File: src/ucp/core/ucp_worker.c [hunk 1]
```diff
@@ -1090,6 +1090,69 @@ ucp_worker_get_sys_device_distance(ucp_context_h context,
...
+
+    if ((context->config.ext.rndv_mode == UCP_RNDV_MODE_AUTO) ||
+    (context->config.ext.rndv_mode == UCP_RNDV_MODE_GET_ZCOPY) ||
+    (context->config.ext.rndv_mode == UCP_RNDV_MODE_GET_PIPELINE) ||
+    (context->config.ext.rndv_mode == UCP_RNDV_MODE_RKEY_PTR)) {
+        iface_params->features |= UCT_IFACE_FEATURE_GET;
+    }
+
+    if ((context->config.ext.rndv_mode == UCP_RNDV_MODE_PUT_ZCOPY) ||
+         (context->config.ext.rndv_mode == UCP_RNDV_MODE_PUT_PIPELINE) ||
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:1147]: pipeline protocols need both get and put for memtype copy
so if have gpu memory both put and get have to be enabled

File: src/uct/base/uct_iface.c [hunk 2]
```diff
@@ -523,6 +523,12 @@ UCS_CLASS_INIT_FUNC(uct_base_iface_t, uct_iface_ops_t *ops,
                                                     ERR_HANDLER_FLAGS, 0);
     self->err_handler_arg   = UCT_IFACE_PARAM_VALUE(params, err_handler_arg,
                                                     ERR_HANDLER_ARG, NULL);
+
+    /* For backward compatibility, in order to not break existing code -
+       if no features are explicitly specified, initialize all features support */
+    self->features          = UCT_IFACE_PARAM_VALUE(params, features, FEATURES,
+                                                    UCT_IFACE_FEATURE_AM | UCT_IFACE_FEATURE_PUT | UCT_IFACE_FEATURE_GET |
+                                                    UCT_IFACE_FEATURE_AMO32 | UCT_IFACE_FEATURE_AMO64 | UCT_IFACE_FEATURE_TAG);
```
Comment (yosefe) [inline src/uct/base/uct_iface.c:531]: line too long

File: src/uct/base/uct_iface.h [hunk 3]
```diff
@@ -209,6 +209,35 @@ enum {
...
+
+/**
+ * @brief Add a capability to interface attributes.
+ *
+ * @param _iface_attr Interface attributes to add capability to.
+ * @param _cap_name   Capability to add.
+ *
+ */
+#define UCT_IFACE_ADD_CAP(_iface_attr, _cap_name) \
+    { \
```
Comment (yosefe) [inline src/uct/base/uct_iface.h:236]: still needed?

File: src/uct/ib/dc/dc_mlx5.c [hunk 4]
```diff
@@ -404,6 +404,12 @@ static int UCS_F_ALWAYS_INLINE
 uct_dc_mlx5_force_full_handshake(uct_dc_mlx5_iface_t *iface, int full_handshake_config)
 {
     if (full_handshake_config == UCS_CONFIG_AUTO) {
+        /* If RDMA_WRITE is not needed, can use half-handshake */
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:407]: in addition to this: if uct_ib_mlx5_iface_has_ar(), and full handshake is not enabled (for whatever reason), the iface should remove PUT capabilities from uct_iface_query()
otherwise seems nothing prevents UCP from doing PUT even if did not pass this as a needed feature

File: test/gtest/ucp/test_ucp_worker.cc [hunk 5]
```diff
@@ -801,6 +801,33 @@ class test_ucp_modify_uct_cfg : public test_ucp_context {
...
+        return (get_variant_value() == UCP_RNDV_MODE_PUT_ZCOPY);
+    }
+
+    static void get_test_variants(std::vector<ucp_test_variant> &variants)
+    {
+        add_variant_with_value(variants, UCP_FEATURE_AM,
+                               UCP_RNDV_MODE_GET_ZCOPY, "get_zcopy");
+        add_variant_with_value(variants, UCP_FEATURE_AM,
+                               UCP_RNDV_MODE_PUT_ZCOPY, "put_zcopy");
+    }
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_worker.cc:828]: seems the test is incomplete..

File: src/uct/ib/dc/dc_mlx5.c [hunk 6]
```diff
@@ -404,6 +412,12 @@ static int UCS_F_ALWAYS_INLINE
...
+        if (!(iface->super.super.super.super.features &
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:416]: should be something like
```c
return (iface->super.super.super.super.features & 
        UCT_IFACE_FEATURE_PUT) &&
       (read capability bit from hw if supported) &&
       uct_ib_mlx5_iface_has_ar(&iface->super.super.super);
```

File: src/uct/ib/dc/dc_mlx5.c [hunk 7]
```diff
@@ -849,8 +856,7 @@ uct_dc_mlx5_iface_create_dcis(uct_dc_mlx5_iface_t *iface,
         for (i = 0; i < iface->tx.ndci; ++i) {
             status = uct_dc_mlx5_iface_create_dci(
                     iface, pool_index, dci_index, pool_index % num_paths,
-                    uct_dc_mlx5_force_full_handshake(
-                            iface, config->dci_full_handshake));
+                    iface->flags & UCT_DC_MLX5_IFACE_FLAG_DCI_FULL_HANDSHAKE);
```
[the remark this answers is not in the document]
Comment (gleon99) [reply src/uct/ib/dc/dc_mlx5.c:859]: No - it is used also for KA (`uct_dc_mlx5_iface_keepalive_init`).

File: src/ucp/core/ucp_worker.c [hunk 8]
```diff
@@ -1128,22 +1119,23 @@ ucp_worker_fill_uct_features(ucp_context_h context, uct_md_attr_t *md_attr,
...
-    if ((context->config.ext.rndv_mode == UCP_RNDV_MODE_AUTO) ||
-        (context->config.ext.rndv_mode == UCP_RNDV_MODE_GET_ZCOPY) ||
-        (context->config.ext.rndv_mode == UCP_RNDV_MODE_GET_PIPELINE) ||
-        (context->config.ext.rndv_mode == UCP_RNDV_MODE_RKEY_PTR) ||
-        (context->config.ext.rndv_mode == UCP_RNDV_MODE_PUT_PIPELINE && has_gpu_memory)) {
+    if (context->num_mem_type_detect_mds > 0) {
+        iface_params->features |= UCT_IFACE_FEATURE_GET | UCT_IFACE_FEATURE_PUT;
+    }
+
+    if (context->config.ext.rndv_mode == UCP_RNDV_MODE_AUTO ||
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:1126]: use ( ) around == expression

File: src/ucp/core/ucp_worker.c [hunk 9]
```diff
@@ -1128,22 +1119,23 @@ ucp_worker_fill_uct_features(ucp_context_h context, uct_md_attr_t *md_attr,
...
+        context->config.ext.rndv_mode == UCP_RNDV_MODE_GET_ZCOPY ||
+        context->config.ext.rndv_mode == UCP_RNDV_MODE_GET_PIPELINE ||
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:1128]: we don't need to check PIPELINE; pipeline will be used only if have GPU memory, and that is already covered by line 1123. Same for line 1134

File: src/ucp/core/ucp_worker.c [hunk 10]
```diff
@@ -1128,22 +1119,23 @@ ucp_worker_fill_uct_features(ucp_context_h context, uct_md_attr_t *md_attr,
...
+        context->config.ext.rndv_mode == UCP_RNDV_MODE_RKEY_PTR) {
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:1129]: why RKEY_PTR required FEATURE_GET?

File: src/uct/ib/dc/dc_mlx5.c [hunk 11]
```diff
@@ -199,12 +200,15 @@ static ucs_status_t uct_dc_mlx5_iface_query(uct_iface_h tl_iface, uct_iface_attr
...
 
-    if (uct_ib_mlx5_iface_has_ar(&iface->super.super.super) &&
-        (!(iface->flags & (UCT_DC_MLX5_IFACE_FLAG_DCI_FULL_HANDSHAKE |
-                           UCT_DC_MLX5_IFACE_FLAG_KEEPALIVE_FULL_HANDSHAKE |
-                           UCT_DC_MLX5_IFACE_FLAG_DCT_FULL_HANDSHAKE)))) {
-        iface_attr->cap.flags &= (~UCT_IFACE_FLAG_PUT_BCOPY & ~UCT_IFACE_FLAG_PUT_ZCOPY);
+    is_full_handshake = iface->flags &
+                        (UCT_DC_MLX5_IFACE_FLAG_DCI_FULL_HANDSHAKE |
+                         UCT_DC_MLX5_IFACE_FLAG_KEEPALIVE_FULL_HANDSHAKE |
+                         UCT_DC_MLX5_IFACE_FLAG_DCT_FULL_HANDSHAKE);
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:206]: should be 
ucs_test_all_flags(iface->flags, UCT_DC_MLX5_IFACE_FLAG_DCI_FULL_HANDSHAKE|UCT_DC_MLX5_IFACE_FLAG_DCT_FULL_HANDSHAKE)

because:
1. we don't care about keepalive DCI, we don't post PUT ops on it
2. need both DCI and DCT full handshake (or just DCI) to enable PUT

File: src/uct/ib/dc/dc_mlx5.c [hunk 12]
```diff
@@ -199,12 +200,15 @@ static ucs_status_t uct_dc_mlx5_iface_query(uct_iface_h tl_iface, uct_iface_attr
...
 
+    if (uct_ib_mlx5_iface_has_ar(&iface->super.super.super) &&
+        !is_full_handshake) {
+        iface_attr->cap.flags &= (~UCT_IFACE_FLAG_PUT_BCOPY &
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:210]: also remove PUT_SHORT

File: src/uct/ib/dc/dc_mlx5.c [hunk 13]
```diff
@@ -413,12 +417,10 @@ uct_dc_mlx5_force_full_handshake(uct_dc_mlx5_iface_t *iface, int full_handshake_
...
         /* If RDMA_WRITE is not needed, can use half-handshake */
-        if (!(iface->super.super.super.super.features &
-              UCT_IFACE_FEATURE_PUT)) {
-            return 0;
-        }
-
-        return uct_ib_mlx5_iface_has_ar(&iface->super.super.super);
+        return (iface->super.super.super.super.features &
+                UCT_IFACE_FEATURE_PUT) &&
+               (iface->version_flag & UCT_IB_DEVICE_FLAG_DC_V2) &&
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:422]: this is not what we should check..

File: src/uct/ib/dc/dc_mlx5.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5.c:859]: ok

File: src/uct/ib/dc/dc_mlx5.c [hunk 14]
```diff
@@ -1392,18 +1373,15 @@ static UCS_CLASS_INIT_FUNC(uct_dc_mlx5_iface_t, uct_md_h tl_md, uct_worker_h wor
...
-
-    if (uct_dc_mlx5_force_full_handshake(self, config->dci_ka_full_handshake)) {
-        self->flags |= UCT_DC_MLX5_IFACE_FLAG_KEEPALIVE_FULL_HANDSHAKE;
-    }
-
-    if (uct_dc_mlx5_force_full_handshake(self, config->dct_full_handshake)) {
-        self->flags |= UCT_DC_MLX5_IFACE_FLAG_DCT_FULL_HANDSHAKE;
+    if ((params->field_mask & UCT_IFACE_PARAM_FIELD_FEATURES) &&
+        !(params->features & UCT_IFACE_FEATURE_PUT)) {
+            self->flags |= UCT_DC_MLX5_IFACE_FLAG_DISABLE_PUT;
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:1378]: weird indent

File: src/uct/ib/dc/dc_mlx5.h [hunk 15]
```diff
@@ -50,6 +50,18 @@ struct ibv_ravh {
         _txwq = &(_iface)->tx.dcis[_dci].txwq; \
     }
 
+#define UCT_DC_MLX5_CHECK_FORCE_FULL_HANDSHAKE(_self, _config, _config_name, \
+                                               _flag_name, _status, _err) \
+    if (!(_self->version_flag & UCT_IB_DEVICE_FLAG_DC_V2) && \
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.h:55]: use ( ) around macro args usage

File: src/uct/ib/dc/dc_mlx5.h [hunk 16]
```diff
@@ -50,6 +50,18 @@ struct ibv_ravh {
...
+        _config->_config_name##_full_handshake == UCS_YES) { \
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.h:56]: ( ) around == expression

File: src/uct/ib/dc/dc_mlx5.h [hunk 17]
```diff
@@ -50,6 +50,18 @@ struct ibv_ravh {
...
+        _status = UCS_ERR_UNSUPPORTED; \
+        goto _err; \
+    } \
+    if (_config->_config_name##_full_handshake == UCS_YES || \
+        _config->_config_name##_full_handshake == UCS_TRY) { \
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.h:61]: 1. ( ) around == expression
2. can just check that != UCS_NO

File: src/uct/ib/dc/dc_mlx5.h [hunk 18]
```diff
@@ -50,6 +50,18 @@ struct ibv_ravh {
...
+        _self->flags |= UCT_DC_MLX5_IFACE_FLAG_##_flag_name##_FULL_HANDSHAKE; \
+    }
+
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.h:64]: 2 space lines in H file

File: test/gtest/ucp/test_ucp_wireup.cc [hunk 19]
```diff
@@ -1137,8 +1137,8 @@ class test_ucp_wireup_unified : public test_ucp_wireup {
 public:
     static void get_test_variants(std::vector<ucp_test_variant>& variants)
     {
-        add_variant_with_value(variants, UCP_FEATURE_TAG,
-                               TEST_TAG | UNIFIED_MODE, "uni");
+        add_variant_with_value(variants, UCP_FEATURE_RMA,
```
Comment (gleon99) [inline test/gtest/ucp/test_ucp_wireup.cc:1140]: @yosefe  the test failed w/o this change, because:
Only tag feature requested -> RMA is disabled -> PUT explicitly disabled in DC -> DC is not preferred over RC.
Do you think there is a better way to fix this?

File: src/uct/base/uct_iface.c [hunk 20]
```diff
@@ -523,6 +523,16 @@ UCS_CLASS_INIT_FUNC(uct_base_iface_t, uct_iface_ops_t *ops,
                                                     ERR_HANDLER_FLAGS, 0);
     self->err_handler_arg   = UCT_IFACE_PARAM_VALUE(params, err_handler_arg,
                                                     ERR_HANDLER_ARG, NULL);
+
+    /* For backward compatibility, in order to not break existing code -
+       if no features are explicitly specified, initialize all features support */
+    self->features          = UCT_IFACE_PARAM_VALUE(params, features, FEATURES,
```
Comment (yosefe) [inline src/uct/base/uct_iface.c:529]: where is this field used?

File: test/gtest/ucp/test_ucp_wireup.cc (hunk 19 above)
Comment (yosefe) [reply test/gtest/ucp/test_ucp_wireup.cc:1140]: probably need to update ucp_worker_iface_find_better to check only flags that are needed for UCP features, and not all flags, in here:
```c
        /* Check that another iface: */
        if (/* 1. Supports all capabilities of the target iface (at least),
             *    except ...CONNECT_TO... and KEEPALIVE-related caps. */
            ucs_test_all_flags(if_iter->attr.cap.flags, test_flags) &&
```

File: src/ucp/core/ucp_worker.c [hunk 21]
```diff
@@ -803,6 +803,84 @@ static void ucp_worker_uct_iface_close(ucp_worker_iface_t *wiface)
...
+        iface_params->features |= UCT_IFACE_FEATURE_AMO64;
+    }
+
+    if (context->num_mem_type_detect_mds > 0) {
+        iface_params->features |= UCT_IFACE_FEATURE_GET | UCT_IFACE_FEATURE_PUT;
+    }
+
+    if ((context->config.ext.rndv_mode == UCP_RNDV_MODE_AUTO) ||
+        (context->config.ext.rndv_mode == UCP_RNDV_MODE_GET_ZCOPY) ||
+        (context->config.ext.rndv_mode == UCP_RNDV_MODE_RKEY_PTR)) {
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:840]: remove rkey_ptr?

File: src/ucp/core/ucp_worker.c [hunk 22]
```diff
@@ -803,6 +803,84 @@ static void ucp_worker_uct_iface_close(ucp_worker_iface_t *wiface)
...
+
+    if (context->config.ext.rndv_mode == UCP_RNDV_MODE_PUT_ZCOPY) {
+        iface_params->features |= UCT_IFACE_FEATURE_PUT;
+    }
+}
+
+static uint64_t ucp_worker_get_exclude_caps(ucp_worker_h worker)
+{
+    uct_iface_params_t iface_params = {0};
+    uint64_t caps = 0;
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:852]: 1. rename to exclude_caps
2. align =

File: src/ucp/core/ucp_worker.c [hunk 23]
```diff
@@ -803,6 +803,84 @@ static void ucp_worker_uct_iface_close(ucp_worker_iface_t *wiface)
     }
 }
 
+static void ucp_worker_fill_uct_features(ucp_context_h context,
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:806]: maybe return uint64_t (features)?

File: src/ucp/core/ucp_worker.c [hunk 24]
```diff
@@ -803,6 +803,84 @@ static void ucp_worker_uct_iface_close(ucp_worker_iface_t *wiface)
...
+        caps |= UCT_IFACE_FLAG_TAG_EAGER_SHORT |
+                UCT_IFACE_FLAG_TAG_EAGER_BCOPY |
+                UCT_IFACE_FLAG_TAG_EAGER_ZCOPY |
+                UCT_IFACE_FLAG_TAG_RNDV_ZCOPY;
+    }
+
+    if (!(iface_params.features & UCT_IFACE_FEATURE_AM)) {
+        caps |= UCT_IFACE_FLAG_AM_SHORT |
+                UCT_IFACE_FLAG_AM_BCOPY |
+                UCT_IFACE_FLAG_AM_ZCOPY;
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:866]: remove

File: src/ucp/core/ucp_worker.c [hunk 25]
```diff
@@ -824,18 +902,19 @@ static int ucp_worker_iface_find_better(ucp_worker_h worker,
...
     ucp_worker_iface_t *if_iter;
-    uint64_t test_flags;
+    uint64_t test_flags, exclude_caps;
     double latency_iter, latency_cur, bw_cur;
 
     ucs_assert(wiface != NULL);
 
     latency_cur = ucp_tl_iface_latency(ctx, &wiface->attr.latency);
     bw_cur      = ucp_tl_iface_bandwidth(ctx, &wiface->attr.bandwidth);
+    exclude_caps = ucp_worker_get_exclude_caps(worker);
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:912]: align =

File: src/uct/api/uct.h [hunk 26]
```diff
@@ -1160,7 +1160,9 @@ struct uct_iface_params {
 
     /**
      * UCT @ref uct_iface_feature "features" that are used for interface
-     * initialization.
+     * initialization. Specifying a feature doesn't guarantee its availability,
+     * because a transport might not implement it. The default value is the set
+     * of all existing features.
```
Comment (yosefe) [inline src/uct/api/uct.h:1165]: move to another PR

File: src/uct/base/uct_iface.c (hunk 20 above)
Comment (yosefe) [reply src/uct/base/uct_iface.c:529]: reminder

File: src/uct/ib/dc/dc_mlx5.c [hunk 27]
```diff
@@ -199,6 +199,12 @@ static ucs_status_t uct_dc_mlx5_iface_query(uct_iface_h tl_iface, uct_iface_attr
                                    max_am_inline,
                                    UCT_RC_MLX5_TM_EAGER_ZCOPY_MAX_IOV(UCT_IB_MLX5_AV_FULL_SIZE));
 
+    if (iface->flags & UCT_DC_MLX5_IFACE_FLAG_DISABLE_PUT) {
+        iface_attr->cap.flags &= (~UCT_IFACE_FLAG_PUT_SHORT &
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:203]: ~(A|B|C)

File: src/uct/ib/dc/dc_mlx5.c [hunk 28]
```diff
@@ -317,6 +323,7 @@ static ucs_status_t uct_dc_mlx5_iface_create_dci(uct_dc_mlx5_iface_t *iface,
         attr.super.max_inl_cqe[UCT_IB_DIR_RX] = 0;
         attr.uidx           = htonl(dci_index) >> UCT_IB_UIDX_SHIFT;
         attr.full_handshake = full_handshake;
+        attr.rdma_wr_disabled = iface->flags & UCT_DC_MLX5_IFACE_FLAG_DISABLE_PUT;
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:326]: align =

File: src/uct/ib/dc/dc_mlx5.h [hunk 29]
```diff
@@ -50,6 +50,18 @@ struct ibv_ravh {
         _txwq = &(_iface)->tx.dcis[_dci].txwq; \
     }
 
+#define UCT_DC_MLX5_CHECK_FORCE_FULL_HANDSHAKE(_self, _config, _config_name, \
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.h:53]: smth like
/* Set iface flag for enabling full handshake on DCI/DCT according user configuration.
   Fail if forcing full-handlshake on a HW that does not support it. */

File: src/uct/ib/dc/dc_mlx5_devx.c [hunk 30]
```diff
@@ -48,7 +47,9 @@ ucs_status_t uct_dc_mlx5_iface_devx_create_dct(uct_dc_mlx5_iface_t *iface,
     UCT_IB_MLX5DV_SET(dctc, dctc, rre, true);
     UCT_IB_MLX5DV_SET(dctc, dctc, rwe, true);
     UCT_IB_MLX5DV_SET(dctc, dctc, rae, true);
-    UCT_IB_MLX5DV_SET(dctc, dctc, force_full_handshake, !!full_handshake);
+    UCT_IB_MLX5DV_SET(dctc, dctc, force_full_handshake,
+                      !!(iface->flags &
+                         UCT_DC_MLX5_IFACE_FLAG_DCT_FULL_HANDSHAKE));
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_devx.c:52]: maybe helper var

File: src/ucp/core/ucp_worker.c [hunk 31]
```diff
@@ -803,82 +803,74 @@ static void ucp_worker_uct_iface_close(ucp_worker_iface_t *wiface)
...
     }
+
+    return features;
 }
 
 static uint64_t ucp_worker_get_exclude_caps(ucp_worker_h worker)
 {
-    uct_iface_params_t iface_params = {0};
-    uint64_t caps = 0;
+    uint64_t exclude_caps = 0, features;
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:850]: separate 
```
uint64_t features     = ucp_worker_get_uct_features(worker->context);
uint64_t exclude_caps = 0;
```

File: src/ucp/core/ucp_worker.c [hunk 32]
```diff
@@ -1203,6 +1274,9 @@ ucs_status_t ucp_worker_iface_open(ucp_worker_h worker, ucp_rsc_index_t tl_id,
         iface_params->am_alignment    = worker->am.alignment;
     }
 
+    iface_params->field_mask |= UCT_IFACE_PARAM_FIELD_FEATURES;
+    iface_params->features    =  ucp_worker_get_uct_features(worker->context);
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:1278]: one space after =

**Review by yosefe — APPROVED** [review]

**Review by yosefe — APPROVED** [review]
